packages/multi-step/src/accounts-multi-step.ts

```diff
@@ -0,0 +1,126 @@
+import { AuthenticationService, DatabaseInterface, User } from '@accounts/types';
+import { AccountsServer } from '@accounts/server';
+
+import { createServiceId, createToken, validateToken } from './utils/crypto';
+import { StepParams, StepResult, MultiStepLoginParams, ErrorMessages } from './types';
+import { errors } from './errors';
+
+export interface AccountsMultiStepOptions {
+  errors: ErrorMessages;
+}
+
+interface DBService {
+  id: string;
+  nextStep?: number;
+  hashedToken?: string;
+}
+
+type AuthenticationServiceArray = {
+  0: AuthenticationService;
+  1: AuthenticationService;
+} & AuthenticationService[];
+
+const defaultOptions = {
+  errors,
+};
+
+export default class MultiStepAuthenticationService implements AuthenticationService {
+  public serviceName = 'multi-step-authentication';
+  public server!: AccountsServer;
+  private db!: DatabaseInterface;
+  private steps: AuthenticationServiceArray;
+  private options: AccountsMultiStepOptions & typeof defaultOptions;
+
+  constructor(steps: AuthenticationServiceArray, options?: AccountsMultiStepOptions) {
+    this.options = { ...defaultOptions, ...options };
+
+    if (steps.length < 2) {
+      throw new Error(this.options.errors.notEnoughAuthenticationServices);
+    }
+
+    this.steps = steps;
+  }
+
+  public setStore(store: DatabaseInterface) {
+    this.db = store;
+  }
+
+  public async authenticateStep({ index, serviceId, ...params }: StepParams): Promise<StepResult> {
```

Comment:
@ozsay Not sure I understand how this works. Could you write a small flow summary in a comment?
Who calls `authenticateStep`?

Reply:
currently, it's very much abstract due to changes required for us in `graphql-api`.

But the idea is that the client will call it.

let's say we want to make a two-factor authentication with username-password and code received from SMS.

so the flow is like this (considering the changes here #776):

1. client will call `authenticateStep({index: 0, username: 'abc', password: '123'})` and will receive: `{ nextStep: 1, serviceId: '1234' }`
2. client will call `prepareAuthentication` from the code service.
3. after he got the SMS he can authenticate the next step:  `authenticateStep({index: 1, code: '4455', serviceId: '1234'})` and will receive the loginToken: `{serviceId: '1234', token: '...'}`
4. then he can perform the regular `loginWithService`.

Reply:
Ah okay so this way we the client is actually oblivious to how many authentication steps there are? is there something in the response that indicates to the client that he should activate sms rather then something like an authenticator app?

Reply:
Currently, the client should be aware of the authentication steps since I can't predict here the flow of the authentication (can involve SMS sending, EMAIL...)